src/resources/dart-3-migration.md

```diff
@@ -0,0 +1,494 @@
+
+* The following declarations can no longer be implemented or extended:
+
+  - `MapEntry`
+  - `OutOfMemoryError`
+  - `StackOverflowError`
+  - `Expando`
+  - `WeakReference`
+  - `Finalizer`
+
+The MapEntry value class is restricted to enable later optimizations. The remaining classes are tightly coupled to the platform and not intended to be subclassed or implemented.
+
+#### `dart:collection`
+
+* The following interface can no longer be extended, only implemented:
+
+  - `Queue`
+
+* The following implementation classes can no longer be implemented:
+  - `LinkedList`
+  - `LinkedListEntry`
+
+* The following implementation classes can no longer be implemented
+  or extended:
+  - `HasNextIterator` (Also deprecated.)
+  - `HashMap`
+  - `LinkedHashMap`
+  - `HashSet`
+  - `LinkedHashSet`
+  - `DoubleLinkedQueue`
+  - `ListQueue`
+  - `SplayTreeMap`
+  - `SplayTreeSet`
+
+## Dart 3 tools changes
+
+### Removed tools
+
+Historically the Dart team has offered a number of smaller developer tools for
+things like formatting code (`dartfmt`), analyzing code (`dartanalyzer`), etc.
+In Dart 2.10 (October 2020) we introduced a new unified Dart developer tool, the
+[`dart` tool](https://dart.dev/tools/dart-tool).
+
+#### Scope
+
+This is an *unversioned* change, that applies to all Dart 3 code.
+
+#### Symptom
+
+In Dart 3 these smaller tools do not exist, and have been replaced by the new
+combined `dart` tool.
+
+#### Migration
+
+Use new sub-commands available in the `dart` tool:
+
+| Historical tool | `dart` replacement                                       | Deprecation | Discontinuation |
+|-----------------|----------------------------------------------------------|-------------|-----------------|
+| `stagehand`     |  `dart create`                                           | [2.14](https://github.com/dart-lang/stagehand/issues/671)        | 2.14*  |
+| `dartfmt`       | [`dart format`](https://dart.dev/tools/dart-format)      | [2.14](https://github.com/dart-lang/dart_style/issues/986)        | [2.15](https://github.com/dart-lang/dart_style/issues/986)            |
+| `dart2native`   | [`dart compile exe`](https://dart.dev/tools/dart-compile)    | [2.14](https://github.com/dart-lang/sdk/commit/cac00e9d956a6f7ef28628989912d971f6b908d4)        | [2.15](https://github.com/dart-lang/sdk/commit/6c5fb84716b1f257b170351efe8096fe2af2809b)            |
+| `dart2js`       | [`dart compile js`](https://dart.dev/tools/dart-compile)    | [2.17](https://github.com/dart-lang/sdk/commit/8415b70e75b1d5bbe8251fa6a9eab2d970cf9eec)         | [2.18](https://github.com/dart-lang/sdk/commit/69249df50bcc7a0489176efd3fd79fff018f1b91)             |
+| `dartdevc`      | none                              | [2.17](https://github.com/dart-lang/sdk/commit/5173fd2d224f669fd8d0a1d21adbfd6187d10f53)         | [2.18](https://github.com/dart-lang/sdk/commit/69249df50bcc7a0489176efd3fd79fff018f1b91)             |
```

Comment:
Wait, really? We don't support DDC anymore? I thought that `dart compile js` just picked between DDC functionality and dart2js functionality based on whether the user is asking for a debug build, or something.

If I'm right about this, it would be better for this line to say `dart compile js` instead of "none", maybe with some explanatory text afterwards to clarify how `dart compile js` figures out which mode you want.

If I'm wrong, and we really don't support DDC anymore, it would be good to have some more text here, explaining why we've decided not to support it anymore, and giving recommendations to users who were relying on that functionality.

Reply:
I'm a bit in doubt... @kevmoo can you keep me honest here?

Reply:
DDC was never really supported...directly. Invoking it has always been *crazy*. We tell folks to use `webdev` or `pub run build_runner serve` etc

Reply:
Ah, interesting. In that case, I think a little bit of explanatory text would be nice, just so that folks don't panic and think that we're taking DDC away 😃

Reply:
Linking to `webdev` now